src/parser.js

```diff
@@ -34,99 +46,171 @@ class Parser {
+          // Add escaped single quotes (\') to attribute values
+          const val = value.replace(/'/g, "\\'")
+          if (!val) {
+            attributes.push(name)
+          } else {
+            attributes.push(`${name}='${val}'`)
+          }
           break
+        }
       }
+    }
 
-      // Remove div tagName
-      if (tagName === 'div' && shorten) {
-        pugNode = pugNode.replace('div', '')
-      }
-    })
-
-    if (attributeList.length) {
-      pugNode += `(${attributeList.join(', ')})`
+    if (attributes.length) {
+      pugNode += `(${attributes.join(', ')})`
     }
 
     return pugNode
   }
 
-  isUniqueTextNode (node) {
-    return node.childNodes.length === 1 && isTextNode(node.childNodes[0])
+  /**
+   * formatPugNode applies the correct indent for the current line,
+   * and formats the value as either as a single or multiline string.
+   *
+   * @param {String} node - The pug node (e.g. header(class='foo'))
+   * @param {String} value - The node's value
+   * @param {Number} level - Current tree level to generate indent
+   * @param {String} blockChar - The character used to denote a multiline value
+   */
+  formatPugNode(node, value = '', level, blockChar = '.') {
+    const indent = this.getIndent(level)
+    const result = `${indent}${node}`
+
+    const lines = value.split('\n')
+
+    if (lines.length > 1) {
+      const indentChild = this.getIndent(level + 1)
+      const multiline = lines.map(line => `${indentChild}${line}`).join('\n')
+      return `${result}${blockChar}\n${multiline}`
+    }
+
+    return value.length ? `${result} ${value}` : result
+  }
+
+  /**
+   * createDoctype formats a #documentType element
+   */
+  createDoctype(node, level) {
+    const indent = this.getIndent(level)
+    return `${indent}doctype html`
+  }
+
+  /**
+   * createComment formats a #comment element.
+   *
+   * Block comments in Pug don't require the dot '.' character.
+   */
+  createComment(node, level) {
+    return this.formatPugNode(COMMENT_NODE_PUG, node.data, level, '')
+  }
+
+  /**
+   * createText formats a #text element.
+   *
+   * A #text element containing only line breaks (\n) indicates
+   * unnecessary whitespace between elements that should be removed.
+   *
+   * Actual text in a single #text element has no significant
+   * whitespace and should be treated as inline text.
+   */
+  createText(node, level) {
+    const { value } = node
+    const indent = this.getIndent(level)
+
+    // Omit line breaks between HTML elements
+    if (/^[\n]+$/.test(value)) {
+      return false
+    }
+
+    return `${indent}| ${value}`
+  }
+
+  /**
+   * createElement formats a generic HTML element.
+   */
+  createElement(node, level) {
+    let pugNode = Parser.getNodeWithAttributes(node)
+
+    if (hasSingleTextNodeChild(node)) {
+      const value = node.childNodes[0].value
+      return this.formatPugNode(pugNode, value, level)
+    }
+
+    return this.formatPugNode(pugNode, node.value, level)
```

Comment:
This could be something like:
```
const value = hasSingleTextNodeChild(node)
  ? node.childNodes[0].value
  : node.value

return this.formatPugNode(pugNode, value, level)
```
i am not a big fan of ternary statements but you could make it easier to read

Reply:
I'm not a big fan either, so I tend to avoid them wherever possible. Makes it much more clearer here, thanks for the tip!